### ISSUE TITLE
[Feature Request] Move `_ObjectImpl` to use `Variant` instead of `!kgen.variant`

### ISSUE THREAD

**Issue by soraros** [modularml/mojo#2167]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

As title.

### What is your motivation for this change?

`Variant` wraps `!kgen.variant` if I understand correctly, move to it makes the code look more regular.

### Any other details?

N/

**Comment by StandinKP** [comment]
Makes sense. Can I work on this?

**Comment by soraros** [comment]
@StandinKP I'm not sure why `_ObjectImpl` is written this way, there might be some good reasons. Let's wait for the team for direction anyways.

**Comment by StandinKP** [comment]
Sure

**Comment by StandinKP** [comment]
@JoeLoser can we have direction here as well? :)

**Comment by soraros** [comment]
@StandinKP I guess you can just go ahead.

**Comment by JoeLoser** [comment]
Sorry, I missed the notification on this. I don't see any reason not to use Variant which already wraps the lower level MLIR type. Feel free to work on this and I'd be happy to accept a PR.

**Comment by StandinKP** [comment]
Great I will get going with this

**Comment by StandinKP** [comment]
@soraros while using the Variant, 
Int64 and Float64 are not register_passable. That's why it is giving me an error that all members must be register passable. Should we add register_passable to them? I think that won't be correct. Still confirming

**Comment by soraros** [comment]
@StandinKP That is odd: they are be `register_passable`, and `Variant` should work with mem type. Could you provide a repro maybe in the form of a draft PR or a gist, so people can take a look?

**Comment by StandinKP** [comment]
Yes @soraros I have raised a draft PR. Please take a look and let me know if I used the Variant wrong

**Comment by soraros** [comment]
@StandinKP Ah, I see the problem: `Variant` is not marked as register passable, while the type wrapping it `_ObjectImpl` is. @JoeLoser Do we simply remove the `@register_passable` on `_ObjectImpl`?

**Comment by StandinKP** [comment]
I tried that but it causes further issues in other fns and also then we will need to add the __move_init__ to _ObjectImpl

**Comment by JoeLoser** [comment]
My first reaction would be to continue down the path of removing the reg-passability pieces from all of it (yes, I know `Variant` itself doesn't have the optimization where it's reg-passable if all its Ts are).  Where did thins break down exactly going down this route?

**Comment by JoeLoser** [comment]
If you go down this route, you shouldn't need the direct `kgen.variant.create` calls: those would just be replaced with calling the constructor of the higher-level `Variant` type in the stdlib.  We probably don't need to manage the int variant indices ourselves either and so on.

**Comment by JoeLoser** [comment]
FYI @Dan13llljws as we were talking about this earlier today.

**Comment by JoeLoser** [comment]
> @JoeLoser how was this closed? Was this done internally?

Yeah, looks like @Dan13llljws worked on this yesterday when I was OOO and landed it just now.  I wish Linear bot mentioned that an internal commit (with its automation) is why the internal Linear issue tracking this public GitHub issue was the reason it was closed.

**Comment by StandinKP** [comment]
Ah got it. Let me know when are you going to release more issues? Thanks
